design-document.md

```diff
@@ -30,6 +30,62 @@ practical.]]
 
 ## System constraints
 
+Prio has inherent constraints derived from the tradeoff between privacy
+guarantees and computational complexity. These tradeoffs influence how
+applications may choose to utilize services implementing the specification.
+
+### Data types have limited resolution due to input-validation complexity
+
+Privacy comes at the cost of computational complexity. While affine-aggregatable
+encodings can compute many useful statistics, they require more bandwidth and
+CPU cycles to account for finite-field arithmetic during input-validation. The
+throughput of the system bounds the volume of data processed, which is
+proportional to the verification circuit's size and the rate of verifications
+performed by each work unit in an aggregator.
+
+Applications that utilize proofs with a large number of gates or a high
+frequency of inputs may need to limit inputs into the system to meet bandwidth
+or compute constraints. Some methods of overcoming these limitations include
+choosing a better representation for the data or introducing sampling into the
+data collection methodology.
+
+### Aggregation services may need to fulfill near-real-time constraints
+
+A near-real-time system must produce a response within a deadline. This
+constraint may be relevant when the value of an aggregate decreases over time.
+
+An example of a near-real-time constraint is the expectation that a can be
+verified and aggregated in a period equal to data collection, given some
+computational budget. Meeting these deadlines may require efficient
+implementations of the input-validation protocol. Applications might batch
+requests and utilize binary serialization to improve throughput.
```

Comment:
What's meant by "binary serialization"?

Reply:
The example I had in mind was using something like Avro over JSON+base64 in cases where string parsing takes up more time than it should.